File: algorithms/[134]加油站/solution.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
const canCompleteCircuit = function(gas, cost) {
  const len = gas.length;

  for (let i = 0; i < len; i += 1) {
    let [gasSum, costSum] = [0, 0];
    let flag = true;
    let next = i;
    do {
      gasSum += gas[next];
      costSum += cost[next];
      if (gasSum < costSum) {
        flag = false;
        break;
      }
      next = next === len - 1 ? 0 : next + 1;
    }
    while (next !== i);

    if (flag) {
      return i;
    }
  }

  return -1;
};

File: algorithms/[134]加油站/solution.py
from typing import List


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        n = len(gas)

        # 模拟从各个加油站出发的情况 time O(n^2)
        for i in range(n):
            if gas[i] >= cost[i]:
                left = 0
                flag = True
                for j in range(i, i + n):
                    left += gas[j % n] - cost[j % n]
                    if left < 0:
                        flag = False
                        break
                if flag:
                    return i

        return -1
